common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,311 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package s3store
+
+import (
+	"bytes"
+	"context"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/blobstore"
+	"github.com/uber/cadence/common/blobstore/blob"
+	"go.uber.org/multierr"
+	"io/ioutil"
+	"math"
+	"net/url"
+)
+
+var (
+	// ErrCheckBucketExists could not verify that bucket directory exists
+	ErrCheckBucketExists = &shared.BadRequestError{Message: "could not verify that bucket directory exists"}
+	// ErrWriteFile could not write file
+	ErrWriteFile = &shared.BadRequestError{Message: "could not write file"}
+	// ErrReadFile could not read file
+	ErrReadFile = &shared.BadRequestError{Message: "could not read file"}
+	// ErrCheckFileExists could not check if file exists
+	ErrCheckFileExists = &shared.BadRequestError{Message: "could not check if file exists"}
+	// ErrDeleteFile could not delete file
+	ErrDeleteFile = &shared.BadRequestError{Message: "could not delete file"}
+	// ErrListFiles could not list files
+	ErrListFiles = &shared.BadRequestError{Message: "could not list files"}
+	// ErrConstructKey could not construct key
+	ErrConstructKey = &shared.BadRequestError{Message: "could not construct key"}
+	// ErrBucketConfigDeserialization bucket config could not be deserialized
+	ErrBucketConfigDeserialization = &shared.BadRequestError{Message: "bucket config could not be deserialized"}
+)
+
+type client struct {
+	s3cli s3iface.S3API
+}
+
+// NewClient returns a new Client backed by file system
+func NewClient(cfg *Config) (blobstore.Client, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region: aws.String(cfg.Region),
+	}))
+
+	s3cli := s3.New(sess)
+	return &client{
+		s3cli: s3cli,
+	}, nil
+}
+func s3upload(ctx context.Context, s3api s3iface.S3API, bucket string, key string, b []byte, tags url.Values) error {
+
+	_, err := s3api.PutObjectWithContext(ctx, &s3.PutObjectInput{
+		Bucket:  aws.String(bucket),
+		Key:     aws.String(key),
+		Body:    bytes.NewReader(b),
+		Tagging: aws.String(tags.Encode()),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return blobstore.ErrBucketNotExists
+			}
+
+		}
+		return err
+	}
+	return nil
+}
+
+func s3download(ctx context.Context, s3api s3iface.S3API, bucket string, key string) ([]byte, error) {
```

Comment:
For all of these methods which take a context can we include a call to something to ensure there is a context timeout something like 

```
func (t *terrablobBlobstore) ensureContextTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
	if _, ok := ctx.Deadline(); ok {
		return ctx, func() {}
	}
	return context.WithTimeout(ctx, defaultBlobstoreTimeout)
}
````

Reply:
Should I add this to all implemented methods in the beginning?

Reply:
Fixed